### PR TITLE
changes in user and group pick list

### PR DESCRIPTION
**Pull request by bodo1987**
This should solve the issues #67, it least it works quit fine for me. Just try it out and merge it if you are happy with it. 

Cheers,
bodo

### CODE REVIEW

**Comment by v1r0x** [comment]
thanks for your PR!
I'll have a look at it and merge it or let you know what's missing. For now: Maybe you could add a check if the current user is an admin and thus add all users.

**Comment by bodo1987** [comment]
good idea!
done!

Cheers,
Bodo

**Comment by bodo1987** [comment]
I just found #50 and implemented the search functions for users and groups. Even if I am not sure if all problems regarding the random population of the pick list are solved (I couldn't reprdouce this issue), it should still provide a better way of navigating through a large pick list now.

RFC

Cheers,
Bodo

**Comment by v1r0x** [comment]
I just tested your code and looks good so far! One last thing: On my server I use rather cryptic user names. Would be great if the search would search for user names **and** display names. I think that's the way the sharing app does. Could you add this?

**Comment by stonerl** [comment]
I'd like to add my 2ct. The list should also contain the display names instead of the user names when available.

**Comment by bodo1987** [comment]
OK - "display names" are displayed as well. Maybe, this would look better when the names in brackets are displayed in a slightly different color (in grey) and the UID in bold letters.

But this has to be done by somebody else. My css skills are negligible ;).

**Comment by v1r0x** [comment]
I'd recommend to only show display name and show user name on hover as `alt`.

**Comment by bodo1987** [comment]
OK OK ... not really what u suggested but I think it is also a nice solution 

And I finally learned something about CSS 

RFC
Bodo

File: templates/create.tmpl.php [hunk 1]
```diff
@@ -108,10 +108,10 @@
 			sort($users, SORT_NATURAL | SORT_FLAG_CASE );
                         foreach ($users as $user) : ?>
                             <li class="cl_user_item cl_access_item" id="user_<?php p($user); ?>" >
-				<?php p($user); ?>
-				(
 				<?php p(OC_User::getDisplayName($user)); ?>
```
Comment (v1r0x) [inline templates/create.tmpl.php]: `OC_User::getDisplayName($user)` is [deprecated](https://doc.owncloud.org/api/classes/OCP.User.html#method_getDisplayName). You should use `$userMgr->get($user)->getDisplayName()` instead.

**Comment by bodo1987** [comment]
OK done

**Comment by v1r0x** [comment]
Thanks for the changes!
I'm not completely happy with the current solution. We should either always display the real user name or only display it on hover or something like that.
Instead of
```php
<?php p($user); ?>
```
I'd propose something like this:
```php
<span id="sec_name"><?php p($user); ?></span>
```
and in the css something like this
```css
#sec_name {
    /*remove the display: hidden; value*/
    Element {
    font-style: italic;
    color: gray;
    font-size: 0.9em;
}
```

**Comment by bodo1987** [comment]
I'm not so sure about this. Currently think about two cases:
The first one, somebody doesn't care about diisplay names. This would lead to same user and display names and whatever he search for he will find the correct user.
The other case is that somebody has numbered or random user names in the style user1/user2/ ... . In this case he probably want to search or pick the real names which are the display names. However, it may be the case that he only have an account ID – for some debug reasons or so. For this rare cases, he will search for the UserID and only under this circumstances it makes sense to show it. But in the normal (search for display name) case, he should not be bothered by cryptic names in the list.

**Comment by v1r0x** [comment]
I think it is almost impossible to find a solution which fits all users. I came up with two new ideas.
1. Add a toggle to show display or user names. Imho this is too much effort for a not complete perfect solution
2. Use display name and use username as fallback if no display name is set. This is how it is currently done on the voting page. Imho the best solution/trade off for now. We should merge this as a first solution and maybe open a new issue to find a better solution. But with that we could have a working search functionality and users can report what the think about it.

What do you think?

**Comment by bodo1987** [comment]
By default, the display name is the same as the user ID if you create a new user. Furthermore, I am not sure if it is possible to have no display name? For me, it doesn't work to delete the whole display name. If it is really not possible, then I wouldn’t see the need to check for empty fields. Besides, it would be relatively complex to check this and to handle all the exceptions.
My idea is to keep the current solution. At least if we can't find any use cases where it would be disadvantageous.

**Comment by v1r0x** [comment]
I thought the `getDisplayName()` returns `null` if the display name is not set, but it returns the username, so you are right: no need to check it.
I'm not that happy with the current solution, because it adds/removes names "randomly". And I think that's one reason why OC has display names, so I would use this solution. And that's the way I did it somewhere else in the code.

**Comment by bodo1987** [comment]
What do you mean by add/remove names "randomly"?  Can you give an example for this - maybe I have still a bug in the implementation. 

Besides this issue I am sorry, but I can't really follow your argumentation. 
I use the display names as default for the pick list - as you requested. However, the user is also free to search for a UserID. If he is doing so, then the userID is displayed in brackets. In my point of view a fair solution. 

And how do you manage to delete the whole display name? It doesn't work via GUI, at least not in my installation.

**Comment by bodo1987** [comment]
OK - I checked again and you were right. I could find some cases where the UID was still displayed even if the search sting didn't match the UID anymore. 
Now the behavior should really be as described already above. 
By default only the display name is displayed.
Only if the search sting matches the UID, the latter is displayed as well.

Cheers,
Bodo

**Comment by v1r0x** [comment]
Didn't talk about a bug, but nice that it was helpful finding one :wink: 
What I meant with "randomly" was the intended behaviour. Displaying the user name if the search term matches doesn't give the user any new information. The user can imagine that (if the search term doesn't match the display name) it matches the real user name. Displaying and removing the user name while typing may detract the user instead giving helpful information.

**Comment by bodo1987** [comment]
Tonight, I got another idea. I could check if the UID and the display name are the same. If it is so, then the UID will never be displayed. So, we could remove the number of UIDs displayed and also, they would be limited to cases where they really add some information.
I would also like to get another opinion - what do you prefer stonerl?

**Comment by stonerl** [comment]
@bodo1987 The standard behavior in ownCloud is to only show the display name. Since ownCloud does not allow empty display names there are only 2 cases to take care of; Either the username is equal to the display name or the username and display name is different.

When you share a file in ownCloud you can either enter the display name or the user name, but the UI will always show the display name only. So I would vote for this behavior since this is what users might expect...

**Comment by bodo1987** [comment]
OK it would be most easy to implement it like this. I did it so far in my way in order to NOT confuse the user. I considered it misleading if you find a name which doesn't match at all to the search string.

However, I change it to the proposed behavior.

**Comment by bodo1987** [comment]
done

I still like the previous version more. But I would say it is up to v1r0x to decide which version to merge.

Cheers,
Bodo

**Comment by v1r0x** [comment]
Thanks for all the effort @bodo1987 really appreciate your help! and thanks for your thoughts @stonerl . I like the new version more.
One last thing: Could you squash some of the commits? E.g. the last 6 commits (starting from ea42600 to 03804ea). Thanks!

**Comment by bodo1987** [comment]
done

**Comment by v1r0x** [comment]
Great, thanks!

**Pull request merged**